### DISCUSSION TITLE
ADSR Class [for uiCustomComponent]: NEED 'OnMouseDown() Snap MouseXY to NodeXY': Any Ctrlr Hooks(or trix) for this?

### DISCUSSION THREAD

**Discussion by unityconstruct** [RomanKubiak/ctrlr#624] [Q&A] [answered]
Working on an ADSR Class and needing `setMousePosition()` type call that snaps the MouseXY to the closest handle.

### To break down what's needed in simple terms:
1. user moves mouse into uiCustomControl
2. mouse down triggers MouseDown() callback -> OnMouseDown( comp, event )
3. determine best target `NODE` & get its `coords`
4. set MOUSE XY = NODE XY ( effectively moving the mouse cursor to the object )
  - the `event` object normally provides `source` 
  - `source` normally has a hook for interacting with the `mouse` object
  - this allows `mouse` XY to be changed programmatically: ie: `event.source:setMousePosition()`

## Any thoughts regarding how one might implement 'snap mouse cursor to XY' are much appreciated
![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/c4d32927-67d3-4e83-8d80-21cb379e983a)

**Comment by damiensellier** [comment]
Hi, there's a JUCE class especially for ADSR management. 
https://docs.juce.com/master/classADSR.html
I'm not sure it's implemented in CTRLR. That would definitely be a shortcut to shaping Amplitude envelopes.
https://forum.juce.com/t/juce-adsr-with-slope/47535

**Reply by damiensellier** [reply]
EDIT: The JUCE ADSR Class won't work, it's audio related... sorry, I have no idea.

If we need new JUCE callbacks attached to the customComponent for that particular task, let me know, I'll add them then.

**Reply by unityconstruct** [reply]
it actually still might have something useable with respect to mousing.. will take a gander. THANK YOU

**Reply by unityconstruct** [reply]
@damiensellier : Found a solution on the juce boards... off to Ctrlr-code to see if there is anything related. Not sure if I can `L()` something to hook into it:

https://forum.juce.com/t/move-the-mouse-to-a-component-solved/23160/2

### Issue:
Is it possible to move the mouse (cursor) programatically to a component? Say I have a component e.g. a cursor that moves “by itself” due to automation. When I click on it (or its containing component), I want the mouse to snap to its position so I can continue record the automation from that moment smoothly without any discontinuity. I.e, the opposite behaviour the juce slider exhibits, where the the thumb snaps to the mouse.

### Solution via C++
Thanks Holy_City. It turned out to be easier than I feared to transfer the drag from the parent component to the sliderThumb.

Here’s the complete solution to move the mouse to the current position of the slider thumb (or whatever appropriate child component) and smoothly continue the drag from its current position (instead of the usual thumb jump to the mouse position)
```cpp
void MyComponent::mouseDown(const MouseEvent& e) noexcept
{
    e.source.setScreenPosition(mySliderThumb->getScreenBounds().getCentre().toFloat());

    MouseEvent eThumb = e.withNewMouseDownPosition(mySliderThumb->getBounds().getCentre());
    mySliderThumb->sliderThumbDragger.startDraggingComponent(mySliderThumb, eThumb);
}
```
and in
```cpp
void MyComponent::mouseDrag(const MouseEvent& e) noexcept
{
    mySliderThumb->sliderThumbDragger.dragComponent(mySliderThumb, e, nullptr);
}
Hmm, well not quite, the mouse cursor isn’t changed to a dragging cursor but stays in its normal arrow mode during the drag, but I gues I can live with that…
```
----
You can call 
```cpp
Component::setMouseCursor (MouseCursor::DraggingHandCursor)
```
 in your mouse event callback.

**Reply by unityconstruct** [reply]
Lua/JuceClasses/LMouseEvent.h doesn't have that hook, but has these... will look more tomorrow

```cpp
  /** Creates a copy of this event with a different position.
      All other members of the event object are the same, but the x and y are
      replaced with these new values.
  */
  MouseEvent withNewPosition (Point<float> newPosition) const noexcept;

  /** Creates a copy of this event with a different position.
      All other members of the event object are the same, but the x and y are
      replaced with these new values.
  */
  MouseEvent withNewPosition (Point<int> newPosition) const noexcept;
```

also tried sniffing with what:
```lua
what(event)
```
```
OnClick_Button saved at 03/24/24 02:06:07
LUA>> Object type [MouseEvent]
-----------------------------------------------------------------

Members:
	         getLengthOfMousePress:	function
	          getMouseDownPosition:	function
	           getMouseDownScreenX:	function
	     getDistanceFromDragStartY:	function
	             getScreenPosition:	function
	               mouseWasClicked:	function
	    getMouseDownScreenPosition:	function
	             getNumberOfClicks:	function
	                    getScreenY:	function
	     getDistanceFromDragStartX:	function
	      getDistanceFromDragStart:	function
	                    getScreenX:	function
	                 getMouseDownY:	function
	                   getPosition:	function
	                 getMouseDownX:	function
	        getOffsetFromDragStart:	function
	           getMouseDownScreenY:	function


Attributes:
	                             1:	string
	                             2:	string
	                             3:	string
	                             4:	string
	                             5:	string
	                             6:	string
```

- the source property of MouseEvent isn't exposed
```lua
what(event.source)
```
```
-----------------------------------------------------------------
Compile: OnClick_Button - OK
LUA>> Object type [nil]
-----------------------------------------------------------------

Members:


Attributes:
```

**Comment by damiensellier** [comment]
Please take a look at @dnaldoog d990 panel, there's a ADSR graph with nodes you can move with mouse down etc : https://sourceforge.net/projects/jd990-super-editor/

![Capture d’écran, le 2024-03-23 à 23 58 47](https://github.com/RomanKubiak/ctrlr/assets/55164385/e3a42069-5a31-4bbd-b5bb-ef9ba06e8617)


https://ctrlr.org/roland-jd990-super-editor/

**Reply by unityconstruct** [reply]
@damiensellier  lol...yessir... that JD990 is what got me started on this 😛 .. I actually have been camping out on it : @dnaldoog .. that 990 is a lovely panel sir!
- sadly, I was sucking at the code extraction and mod, so decided to build from scratch..
  - think what started it all was I was killing Ctrlr while trying to mod the param count/structure ... my EMU adsr structure is a bit different
  - second motivation was that I wanted multiple ADSRs without duplicating all the code for each one
  - third, I was sucking at attempts with color mod'ing
  - I also needed support for envelopes with varying params and negative values(-100 <> 100 )
  - which I think I have worked out.

THANKS for the JUCE info, will poke around & see if I get lucky. If not the 990's drag code code is here looking at me & will probably implement something similar based off of it & just revisit my 'mouse snapping' thing later. I'll update if I find a solution.

<hr/>

## debugging paint callback
- the paint happens so soon that if code is jacked, ctrlr bails before you can do much
- can't debug paint callback without locking oneself out
- so it was a bit of a interesting situation that prevented me from being able to steal code
- no feature request, just context on peculiarities specific unto uiCustomComponent paint dev

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/47fa6d30-a489-44a8-b1a6-7a2afa99a558)

**Reply by dnaldoog** [reply]
> sadly, I was sucking at the code extraction and mod, so decided to build from scratch..

Because the code is so convoluted and overly complicated so that when I went to rewrite it, I decided I would be better to just start from scratch, which  I am gradually doing when I get around to it. But that code has been pretty well tested.

**Comment by Ted-juh** [comment]
What is it exactly that you need to know about mouse movement or dragging? It's not clear to me what the problem is.

**Comment by dobo365** [comment]
Same feeling... Don't understand what the problem is.

In the OnMouseDown method you need to track the anchors zones.
In the Paint method you draw them. You take the values from the buttons/sliders.
You can either use the drag handles or move the buttons/sliders to adjust the env.

For the moment, yep, 3 methods by envelope but you can easily make only 3 for all envelopes if you prefer.
Just a matter to attach the same methods to all envs then to identify the current "comp" and use the corresponding buttons/sliders.
I'm doing that all the time for all my mods with similar behavior and for which I have only 2 methods (OnClick/OnChange).

For the colour you use
	graphics:setColour (Colour(0xFFDDDDDD))
before calling the drawLine, drawRect... functions

According to the envelope type, it is a matter to adapt the standard paint env code that you can find in different panels.

**Reply by Ted-juh** [reply]
Ok, I think I finally get the problem. 

You want to click "somewhere" in the component and make a "node" snap to the point you clicked.

Then you first need to "know" where all the nodes are and do a quick math to calculate the distance to the node that is closer to the point you clicked than the rest of the nodes. 

If you ask me, the best way is to determine the x point of a node (the nodes can't overlap and are in a certain order) and compare that to the x point of the mouse click. Then swap the x,y of the closest node with the ones of the mouse click.

But to make matters worse, you first have to define "which" graph (or colored line) you want to snap as well. Should be possible if you can select the graph that you need. Possibly with a graph (color) selector.

**Comment by unityconstruct** [comment]
@Ted-juh 
yessir, almost😎 ...( & to your point only one graph should be in focus or it would be a hot mess.. )
- don't want to move the node handle to the mouse... but move the mouse to the closest node (snapping)
- needing `mouse snapping` - or in its most basic form.. `moving the mouse programmatically`
- cleaned up the OP for clarity now that I know what I'm looking for
- its actually via `MouseDown()` event, so `MouseDrag()` is not in scope
- 2017 JUCE forum post w/ c++ implementation [Move the mouse to a component](https://forum.juce.com/t/move-the-mouse-to-a-component-solved/23160/5) (c++ code above), but the method doesn't exist in the JUCE 6.0.8 sources at this time
- think I found its new location in 6.0.8: `JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource`
- MouseEvent.source (event.source) yields MouseInputSource
```cpp
     /** The source device that generated this event. */
    MouseInputSource source;
```
- and MouseInputSource.setScreenPosition() moves the mouse
```
    void setScreenPosition (Point<float> p)
    {
        MouseInputSource::setRawMousePosition (ScalingHelpers::scaledScreenPosToUnscaled (p));
    }
```
- think the answer is `not implemented`, but hooks exist in JUCE
- might be as simple as LuaWrap the `event.source` property, but might have to also expose the call `source.setRawMousePosition()`.
- then the call would be something like `event.source:setRawMousePosition(COORDS)`

**Comment by unityconstruct** [comment, marked as answer]
@damiensellier .. I just added LMouseInputSource, built successfully and the object/method are now exposed to Ctrlr. Ctrlr might be lying & just mimicking what's in the `wrapForLua` setup, but about to find out. If it actually works, not sure if you are ok if I pull from CtrlrX, update & do PR. If not, I can just maintain it in my fork. THX for your responses above as they more or less led me to where I needed to be.
- setScreenPosition() takes Point & looks like there's already LPoint, so hopeful its the same Point MIS wants.

# LMouseInputSource Luabind Add

- Source/lua/JuceClasses/LMouseInputSource.h
```cpp
#ifndef L_MOUSE_INPUT_SOURCE
#define L_MOUSE_INPUT_SOURCE

extern  "C"
{
	#include "lua.h"
}

class LMouseInputSource
{
	public:
		static void wrapForLua(lua_State* L);
};

#endif
```


- Source/lua/JuceClasses/LJuce.h
```cpp
#include "LMouseInputSource.h"
```

- Source/lua/JuceClasses/LJuce.cpp
```cpp
LMouseInputSource::wrapForLua(L);
```

- need to FIRST expose the MouseEvent `source` property which provides access to the mouse
- `what(event.source)`, can be used to check if `source` is exposed
- if yes, than it will also show the newly exposed method if successfully added
```cpp
void LMouseEvent::wrapForLua (lua_State *L)
{
	using namespace luabind;

	module(L)
    [
		class_<MouseEvent>("MouseEvent")
			.def("getMouseDownX", &MouseEvent::getMouseDownX)

			.def_readonly("source", &MouseEvent::source)  <<----
			
			.def_readonly("x", &MouseEvent::x)
			.def_readonly("y", &MouseEvent::y)
```

## UPDATE 3/25
- Source/lua/JuceClasses/LCore.cpp
```cpp
void LMouseInputSource::wrapForLua(lua_State* L)
{
	using namespace luabind;

	module(L)
		[
			class_<MouseInputSource>("MouseInputSource")
				.def("setScreenPosition", &MouseInputSource::setScreenPosition)
		];
}
```
## UPDATE 3/25

```lua
--
-- Called when a mouse is down on this component
--
ButtonOnClick = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
what(event.x)
what(event.source)
end
```

```output
LUA>> Object type [number]
-----------------------------------------------------------------

Members:


Attributes:

-----------------------------------------------------------------

LUA>> Object type [MouseInputSource]
-----------------------------------------------------------------

Members:
	             setScreenPosition:	function


Attributes:

-----------------------------------------------------------------


void LMouseEvent::wrapForLua (lua_State *L)
{
	using namespace luabind;

	module(L)
    [
		class_<MouseEvent>("MouseEvent")
			.def("getMouseDownX", &MouseEvent::getMouseDownX)

			.def_readonly("source", &MouseEvent::source)
			
			.def_readonly("x", &MouseEvent::x)
			.def_readonly("y", &MouseEvent::y)
```

**Reply by unityconstruct** [reply]
@damiensellier FYI: Its working
```lua
--
-- Called when a mouse is down on this component
--
ButtonOnClick = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
  what(event.x)
  what(event.source)
  event.source:setScreenPosition(Point(50,50))
  console("move mouse")
end
```

[MouseEventSourceSetScreenPosition.zip](https://github.com/RomanKubiak/ctrlr/files/14737302/MouseEventSourceSetScreenPosition.zip)

**Reply by dnaldoog** [reply]
I was just looking at this but on 5.3.201 - I had got as far as `event.source:setScreenPosition(Point(50,50))` but without the `:`  ` - trying to emulate the C++ from [this post](https://forum.juce.com/t/move-the-mouse-to-a-component-solved/23160/2) but needless to say doesn't work in __5.3.201__,

<pre>
mouseUp = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
local x,y=event.x,event.y
local p=Point()
p:setX(mouseTarget.x)
p:setY(mouseTarget.y)
event.source:setScreenPosition(p)
end
</pre>

Does this panel work on 5.6.31?

### UPDATE EDIT
[Ctrlr Panel MouseChaser_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/14738581/Ctrlr.Panel.MouseChaser_2_0.zip)

**Reply by unityconstruct** [reply]
yeah.. its on my end... tail end of a Ctrlr Marathon apparently.. I didn't capture the luabind for the `event.source`.. posted below, but going to do this the right way & submit PR so there's an accurate diff.. Sorry to waste your time! 

`what(source)` will let us know if `source` got exposed.. & I think it will even show the new set*() method when it reflects the functions...

**Comment by damiensellier** [comment]
Hi UC, sounds good! 
I'll add the new file Source/lua/JuceClasses/LMouseInputSource.h in the next version with the LUA bind lines.
Thanks for your suggestion!

**Reply by damiensellier** [reply]
I just added them and it compiled well on mac. The new class will be available in CtrlrX v5.6.31
Thanks !

**Comment by dnaldoog** [comment]
Thanks @damiensellier for posting that panel as an example of an ADSR (type) graph. 

I have been working on and off on a complete rewrite of the JD-990 panel. The code in that currently released panel was written when I was just starting out with Ctrlr and as such could use a few improvements and streamlining.

Anyway, I have separated the latest graph code and added it to a dedicated panel, which should make it easier to decipher for @unityconstruct and others.

![JD-990 TVF Example panel](https://github.com/RomanKubiak/ctrlr/assets/76831683/943eea88-97c7-4fa1-8846-bf8fbf3aa11d)

[Example TVF graph JD990_1.zip](https://github.com/RomanKubiak/ctrlr/files/14737813/Example.TVF.graph.JD990_1.zip)

**Comment by dnaldoog** [comment]
Bump: does this panel work in 5.6.31?

[Ctrlr Panel MouseChaser_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/14738581/Ctrlr.Panel.MouseChaser_2_0.zip)

**Reply by damiensellier** [reply]
Nope it doesn't work, I get an error on 5.6.30 : 

```
At line [11]: [string "mouseUp"]
Error message: [string "mouseUp"]:11: attempt to index field 'source' (a nil value)
```

**Reply by damiensellier** [reply]
EDIT: The white box is changing with the sliders but when I click on the red area I get the LUA error message as above.

**Reply by dnaldoog** [reply]
I tried it in 5.6.30 On windows exe and it didn’t work, but I thought you only did the luabind in 5.6.31?

**Reply by damiensellier** [reply]
I did put the addons @unityconstruct posted above but don't get it to work. There's probably something missing in UC post.
Source/lua/JuceClasses/LMouseInputSource.h added in projucer
content added in xcode to Source/lua/JuceClasses/LMouseInputSource.h

```
#ifndef L_MOUSE_INPUT_SOURCE
#define L_MOUSE_INPUT_SOURCE

extern  "C"
{
	#include "lua.h"
}

class LMouseInputSource
{
	public:
		static void wrapForLua (lua_State *L);
};

#endif
```

Source/lua/JuceClasses/LJuce.h updated

`#include "LMouseInputSource.h"`

Source/lua/JuceClasses/LJuce.cpp updated

`LMouseInputSource::wrapForLua(L);`

Source/lua/JuceClasses/LCore.cpp updated

```
void LMouseInputSource::wrapForLua(lua_State* L)
{
	using namespace luabind;

	module(L)
		[
			class_<MouseInputSource>("MouseInputSource")
				.def("setScreenPosition", &MouseInputSource::setScreenPosition)
		];
}
```

Am I missing something?

**Reply by Ted-juh** [reply]
Just a quick guess, aren't you missing the constructor as in: 

```
.def(constructor<>())
.def(constructor<const MouseInputSource &>())
```


There might also be a missing enum:

```
.enum ("InputSourceType")
[
value ("mouse", MouseInputSource::mouse),
value ("touch", MouseInputSource::touch),
value ("pen", MouseInputSource::pen)
]
```

**Reply by damiensellier** [reply]
Can you please give me all the details to where I should add those lines? Thanks ;)
I'll try tomorrow. If @unityconstruct got it working I wonder why it's not doing well on my side.

also setScreenPosition() is nowhere to be found in the LUA bind list, maybe it's working with C++ under the hood IDK.

**Reply by damiensellier** [reply]
@Ted-juh I'm trying right now and will let you know...
```

void LMouseInputSource::wrapForLua(lua_State* L)
{
    using namespace luabind;
    
    module(L)
    [
        class_<MouseInputSource>("MouseInputSource")
                .def(constructor<>())
                .def(constructor<const MouseInputSource &>())
                 .enum_("MouseInputSource")
                 [
                  value("mouse", MouseInputSource::mouse),
                  value("touch", MouseInputSource::touch),
                  value("pen", MouseInputSource::pen)
                 ]
                .def("setScreenPosition", &MouseInputSource::setScreenPosition)
     ];
}
```

**Reply by damiensellier** [reply]
Build failed unfortunately.

**Reply by Ted-juh** [reply]
Lol.. both working on the same thing then but I was busy trying to add the whole class.

What is failing, which error?

**Reply by Ted-juh** [reply]
'juce::MouseInputSource': no appropriate default constructor available

**Reply by damiensellier** [reply]
I had like 15 errors in xcode. I tried to remove the constructor with the enum type keeping only the constructor but it failed too.
I find it strange the OP got it working only with those extra lines, we will see...
Here's what I found on the juce forum : https://forum.juce.com/t/accessing-mouseinputsource-from-mouseevent/42763

e.source.setScreenPosition(juce::Point<float>(0,0)); // failing

>>> _e is const, and MouseInputSource::setScreenPosition() is a non-const method. You could use Desktop::getInstance().getMainMouseSource() if you know you’ll only have one input source (on desktop for example) or use ::getNumMouseSources() and ::getMouseSource() to iterate over the sources._

Not sure if setScreenPosition will work in our case.

**Reply by Ted-juh** [reply]
Ah.. dang. It's quite simple. You can't create a new MouseInputSource so the first constructor is at fault. 

```
void LMouseInputSource::wrapForLua(lua_State* L)
{
    using namespace luabind;
    
    module(L)
    [
        class_<MouseInputSource>("MouseInputSource")
                .def(constructor<const MouseInputSource &>())
                 .enum_("MouseInputSource")
                 [
                  value("mouse", MouseInputSource::mouse),
                  value("touch", MouseInputSource::touch),
                  value("pen", MouseInputSource::pen)
                 ]
                .def("setScreenPosition", &MouseInputSource::setScreenPosition)
     ];
}
```

Now it compiles. But not sure if it works.

**Reply by damiensellier** [reply]
Compilation went ok but the 2 demo panels are not working. Time to sleep. Have a nice week!

**Reply by Ted-juh** [reply]
And if you ask me, it should be: 
.def("setScreenPosition", (Point<int> (MouseInputSource::*)() const)&MouseInputSource::setScreenPosition)

See the Juce Class: 
void setScreenPosition (Point< float > newPosition)

**Reply by unityconstruct** [reply]
doh!... @damiensellier 
@Ted-juh nailed it.. forgot to post the one-line add to expose `source`.. it's a Property, so will use _def_readonly() instead of .def()
.. also, I had a helluva time getting the stinking .h file into the project, with Add class wanting to push it to the project root. 
- I gave up and just moved it manually to the juceclasses folder, removed the broken file reference.
- I don't recall if I did [> new.. Existing] or if VS picked up the newly-moved LMouseInputSource.h file automatically.
- I'll pull from master, reapply changes, test & PR to ensure a clean diff
```cpp
.def_readonly("source", &MouseEvent::source)
```

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/c4bdfd6b-e4b9-4baa-8446-10887f852299)

**Comment by unityconstruct** [comment]
- Pulled lastest from master, applied stash on 4 files, built, tested good.
- PR: here: https://github.com/damiensellier/CtrlrX/pull/13, 
- with full diff on changes here: https://github.com/damiensellier/CtrlrX/pull/13/files

**Reply by damiensellier** [reply]
Thanks, I just updated the missing line and it seems to work. From @dnaldoog test panel I have my mouse pointer jumpingto another position when clicking and it returns in the debugger the following : 

```
CtrlrLuaMethodManager::call paintMe
CtrlrLuaMethodManager::call mouseUp
```

With UC panel I get the following message in the LUA console, and the pointer is jumping to the screen pos (x50,y50) as desired:

```
LUA>> Object type [MouseInputSource]
-----------------------------------------------------------------
Members:
	             setScreenPosition:	function
	                        __init:	function
Attributes:
-----------------------------------------------------------------
LUA>> move mouse
```

So, kudos! it seems to work, it will be available in the next release v5.6.31

Have a nice week ;)

**Reply by dnaldoog** [reply]
Wow that’s great work @damiensellier - it’s really exciting to see how quickly you can move on these things and get them incorporated.

just got to find a use for having a jumping mouse now! Ha ha.😂

**Reply by unityconstruct** [reply]
A classic & stylish implementation is to have the MouseDown() event first move the cursor to the close icon before proceeding to honor the Click()

**Reply by dnaldoog** [reply]
It actually is a very good idea now that I think about it, because sometimes centering the mouse on the small rectangles in, for example, a graph can be a little fiddly, whereas if the mouse does a kind of honing in action, I can see that that would be quite cool!

**Comment by Ted-juh** [comment]
I fail to understand why this method would differ from withNewPosition (which is a Mouse event). To me it should pass the new coordinates to the mouse. Doesn't matter that much, it works as intended the way it is now.

User cases: 
Centering on small objects.
Snap to a grid, nodes, basically anything.

Or for a linear bar, snap to the value instead of that the point you clicked becomes the new value.

Or, with a long combobox or menu, make it jump back to the start.

**Reply by dnaldoog** [reply]
https://github.com/RomanKubiak/ctrlr/discussions/619#discussioncomment-8928270

**Reply by Ted-juh** [reply]
So, screen as in screen position means the "monitor" here, I guess. It's not obvious from the description in the mouseInputSource class. Whereas is the Mouse event class it is obvious: 
The coordinates are relative to the top-left of the main monitor.

I wonder what withNewPosition() does because of its description: "All other members of the event object are the same, but the x and y are replaced with these new values.". 

If the component is a member of the mouse event.. its coordinates should correspond to the component and not that of the monitor.

**Reply by dnaldoog** [reply]
Yes, it's all in the name  - screen means the screen - !

Was messing about with:

<pre>
.def("getLocalPoint", (Point<int> (Component::*)(const Component *, Point<int>) const) &Component::getLocalPoint)
//.def("getLocalArea", &Component::getLocalArea)
.def("localPointToGlobal", (Point<int> (Component::*)(Point<int>) const) &Component::localPointToGlobal)
</pre>

but don't seem to be bound?

**Reply by Ted-juh** [reply]
That's in LComponents.cpp ? Yes, it is indeed. Mind that there are two ways to add the getLocalPoint with Luabind. One with <int> and one with <float>. See The component class in Juce. 

How are you using it in a panel?

**Reply by unityconstruct** [reply]
> If the component is a member of the mouse event.. its coordinates should correspond to the component and not that of the monitor.
- absolutely agree. Methinks using manual absolute coords is recipe for disaster due to resolution diffs & monitor sizes. Snapping to component center should safeguard, but I'm even pondering its usefulness for actions in an `LCD display` where smaller icons are possible without driving the user mad trying click on something microscopic.  
- There were a few motivations on snapping - that not surprisingly immediately became moot after working it out, so will revisit once I have this panel happily doing basics - namely making it easier to muck with the envelope handles... 
- I already have 12 envelopes to content with... and still need to add R1L,R2T & R2L for all of these(144 nodes) - snapping should allow for higher information density without increased difficulty mousing stuff.
- can't help but think that the chaotic eye-sore could be simplified with each node consolidating 2 params - though I honesty like having access to one-control-per-param for more detailed tweaks, etc.... And seems snapping might make this widget a bit more survivable.
![2024 03 27 212208 cap](https://github.com/RomanKubiak/ctrlr/assets/40268430/309f3876-f704-4251-9b6f-3495c73a5c73)
- still have LFOs pending that might benefit from UI hacks of some sort
- + there's like 24 `cords` per layer so would be koo to 'snap-drag-snap' cords from SOURCE to DEST for each, using another class that seems similar to messing with ADSRs

now that I figured out that luabind, I'm prob dangerous at this point
